content/BPIP-3.md

```diff
@@ -146,28 +149,113 @@ The following method is added.
+Vouchers, which are ERC721 NFTs, show on marketplaces as soon as they are minted. Currently, vouchers are minted at the same time when an exchange is created, which means that whenever they are resold on a secondary market, the new buyer does not get full dispute period (i.e. transfer does not reset the timers in the protocol). In order to allow buyers on any marketplace to get full dispute period when they are the first owner of the voucher, we need a different solution. 
+
+One possible approach is to create a dedicated bridge contract, where sellers could create NFTs that would show on marketplaces. When buyers on marketplace bought them, this bridge NFT would be burnt and buyers would get redeemable voucher instead. However this approach breaks user flow in some marketplaces and might be questionable since the buyer gets a different NFT than was actually bought.
+
+Another approach was to enable mint on demand directly on voucher contract. Although it does not suffer the same limitations as the previous approach, mint is not a standard ERC721 function, so compatibility with all marketplaces would be hard to achieve.
+
+This leads to this proposal, where vouchers are preminted directly on existing voucher contract. Their existence automatically means they can be transferred as regular ERC721 NFTs, so they can be sold on any marketplace. Compared to the first approach, the voucher here is not burned but simply transferred to a new owner (i.e. first buyer). This first transfer invokes `commitToPremintedOffer` on the protocol, which starts the exchange and effectively converts preminted voucher into a true voucher. After that, the voucher behaves as other vouchers, i.e. dispute period starts at the time of the purchase, the voucher can be redeemed, transferred, etc.
+Important to note here is that we now have two types of vouchers in the voucher contract:
+- preminted vouchers, with no exchange associated and
+- true vouchers, with exactly one exchange associated.
+  
+In current protocol, all vouchers are issued when the protocol invokes `issueVoucher` method. In this proposal, we add an additional method `preMint` which allows the seller (operator) to issue a desired number of vouchers.
+Since currently it holds that the exchange id (in the protocol) always matches the token id (in voucher contract), it's desired that this stays even with premint. To achieve it, we propose a method to reserve range (`reserveRange`), which effectively reserves a desired number of exchange ids in the protocol. Then whenever a preminted voucher is converted into a true voucher, it keeps its token id and gets the matching exchange id in the protocol.
+
+Although range reservation ultimately happens on voucher contracts, the method must be invoked true protocol. That way, it can be ensured that the amount of preminted vouchers never exceeds the quantity available, set in the offer.
```

Comment:
```suggestion
Although range reservation ultimately happens via voucher contracts, the method must be invoked through the protocol. That way, it can be ensured that the amount of preminted vouchers never exceeds the quantity available, set in the offer.
```